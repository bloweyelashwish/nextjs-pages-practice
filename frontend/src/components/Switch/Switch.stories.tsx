import { Meta, StoryObj } from '@storybook/react';
import { expect } from '@storybook/jest';
import { screen, userEvent } from '@storybook/testing-library';

import { Switch } from './Switch';

const meta: Meta<typeof Switch> = {
  tags: ['autodocs'],
  title: 'Controls/Switch',
  component: Switch,
};

export default meta;

type Story = StoryObj<typeof Switch>;

export const BasicSwitch: Story = {
  play: async ({ args }) => {
    await userEvent.click(screen.getByTestId('SwitchVisiblePart'));
    expect(args.onChange).toHaveBeenCalled();
  },
  args: {},
};
