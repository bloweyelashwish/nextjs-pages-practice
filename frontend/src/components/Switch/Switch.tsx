import { FC, ChangeEvent } from 'react';
import styled from '@emotion/styled';

import { boxShadow, transition } from '~/styles/common';
import { useId } from '~/hooks/useId';

const SwitchWrapper = styled.div`
  & input {
    display: none;

    &:checked {
      & ~ label {
        background: ${({ theme }) => theme.components.primary};

        &::after {
          margin-left: 3.5rem;
          background: ${({ theme }) => theme.components.active};
        }
      }
    }
  }
`;
const VisiblePart = styled.label`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  cursor: pointer;
  height: 3rem;
  width: 6rem;
  border-radius: 1.6rem;
  background: ${({ theme }) => theme.components.background};

  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2)};
  ${transition()};

  &::after {
    content: '';
    margin-left: 0.5rem;
    width: 2.1rem;
    height: 2.1rem;
    border-radius: 50%;
    background: ${({ theme }) => theme.components.nonActive};
    ${transition()};
  }
`;

interface SwitchProps {
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
}

export const Switch: FC<SwitchProps> = ({ onChange }) => {
  const fieldId = useId();

  return (
    <SwitchWrapper>
      <input id={fieldId} type="checkbox" onChange={onChange} />
      <VisiblePart htmlFor={fieldId} data-testid="SwitchVisiblePart" />
    </SwitchWrapper>
  );
};
