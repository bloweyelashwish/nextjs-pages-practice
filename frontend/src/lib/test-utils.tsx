import React from 'react';
import type { FC, PropsWithChildren, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';

import { Themes } from '~/styles/themes';

const ThemeWrapper: FC<PropsWithChildren> = ({ children }) => (
  <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
);

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: ThemeWrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
