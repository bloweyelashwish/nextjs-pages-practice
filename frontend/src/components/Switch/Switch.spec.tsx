import userEvent from '@testing-library/user-event';
import { render, screen } from '~/lib/test-utils';

import { Switch } from './Switch';

describe('Switch test cases', () => {
  it('Switch renders', () => {
    const onChange = jest.fn();
    jest.spyOn(Math, 'random').mockReturnValue(0.9999999999);
    const { asFragment } = render(<Switch onChange={onChange} />);

    expect(asFragment()).toMatchSnapshot();
  });
  it('onChange handler is called', async () => {
    const onChange = jest.fn();
    render(<Switch onChange={onChange} />);
    const element = screen.getByTestId('SwitchVisiblePart');

    await userEvent.click(element);

    expect(onChange).toHaveBeenCalled();
  });
});
