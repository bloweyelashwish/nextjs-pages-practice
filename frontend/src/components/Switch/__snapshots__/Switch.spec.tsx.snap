// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch test cases Switch renders 1`] = `
<DocumentFragment>
  <div
    class="css-26o6b"
  >
    <input
      id="prefix-ffffffff920c8"
      type="checkbox"
    />
    <label
      class="css-1hjm6s4"
      data-testid="SwitchVisiblePart"
      for="prefix-ffffffff920c8"
    />
  </div>
</DocumentFragment>
`;
